File: Wire-iOS/Sources/Managers/MediaPlaybackManager.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


@import MediaPlayer;
@import WireSystem;

#import "MediaPlaybackManager.h"
#import "MediaPlayer.h"
#import "AudioTrackPlayer.h"
#import "KeyValueObserver.h"

static NSString* ZMLogTag ZM_UNUSED = @"UI";

NSString *const MediaPlaybackManagerPlayerStateChangedNotification = @"MediaPlaybackManagerPlayerStateChangedNotification";


@interface MediaPlaybackManager ()

@property (nonatomic) id<MediaPlayer> activeMediaPlayer;
@property (nonatomic) AudioTrackPlayer *audioTrackPlayer;

@property (nonatomic, retain) NSObject *titleObserver;

@end

@implementation MediaPlaybackManager

- (instancetype)initWithName:(NSString *)name
{
    self = [super init];
    
    if (self) {
        self.name = name;
        self.audioTrackPlayer = [[AudioTrackPlayer alloc] init];
        self.audioTrackPlayer.mediaPlayerDelegate = self;
        self.titleObserver = nil;
    }
    
    return self;
}

#pragma mark - AVSMedia

@synthesize delegate = _delegate;
@synthesize volume = _volume;
@synthesize recordingMuted = _recordingMuted;
@synthesize looping = _looping;
@synthesize name = _name;

- (void)play
{
    // AUDIO-557 workaround for AVSMediaManager calling play after we say we started to play.
    if (self.activeMediaPlayer.state != MediaPlayerStatePlaying) {
        [self.activeMediaPlayer play];
    }
}

- (void)pause
{
    // AUDIO-557 workaround for AVSMediaManager calling pause after we say we are paused.
    if (self.activeMediaPlayer.state == MediaPlayerStatePlaying) {
        [self.activeMediaPlayer pause];
    }
}

- (void)stop
{
    // AUDIO-557 workaround for AVSMediaManager calling stop after we say we are stopped.
    if (self.activeMediaPlayer.state != MediaPlayerStateCompleted) {
        [self.activeMediaPlayer stop];
    }
}

- (void)resume
{
    [self.activeMediaPlayer play];
}

- (void)reset {
    [self.audioTrackPlayer stop];
    
    self.audioTrackPlayer = [[AudioTrackPlayer alloc] init];
    self.audioTrackPlayer.mediaPlayerDelegate = self;
}

- (BOOL)looping
{
    return NO;
}

- (void)setPlaybackMuted:(BOOL)playbackMuted
{
    if (playbackMuted) {
        [self.activeMediaPlayer pause];
    }
}

- (BOOL)playbackMuted
{
    return NO;
}

#pragma mark - MediaPlayer delegate

- (void)mediaPlayer:(id<MediaPlayer>)mediaPlayer didChangeToState:(MediaPlayerState)state
{
    ZMLogDebug(@"mediaPlayer changed state: %@", @(state));
    [self.changeObserver activeMediaPlayerStateDidChange];

    switch (state) {
        case MediaPlayerStatePlaying:
            if (self.activeMediaPlayer != mediaPlayer) {
                [self.activeMediaPlayer pause];
            }
            
            if ([self.delegate respondsToSelector:@selector(didStartPlayingMedia:)]) {
                [self.delegate didStartPlayingMedia:self];
            }
            
            self.activeMediaPlayer = mediaPlayer;
            [self startObservingMediaPlayerChanges];
            break;
            
        case MediaPlayerStatePaused:
            if ([self.delegate respondsToSelector:@selector(didPausePlayingMedia:)]) {
                [self.delegate didPausePlayingMedia:self];
            }
            break;
            
        case MediaPlayerStateCompleted:
            if (self.activeMediaPlayer == mediaPlayer) {
                self.activeMediaPlayer = nil;
            }

            [self stopObservingMediaPlayerChanges:mediaPlayer];

            if ([self.delegate respondsToSelector:@selector(didFinishPlayingMedia:)]) {
                [self.delegate didFinishPlayingMedia:self]; // this interfers with the audio session
            }
            break;
        case MediaPlayerStateError:
            if ([self.delegate respondsToSelector:@selector(didFinishPlayingMedia:)]) {
                [self.delegate didFinishPlayingMedia:self]; // this interfers with the audio session
            }
            break;
        default:
            break;
    }
    
    [[NSNotificationCenter defaultCenter] postNotificationName:MediaPlaybackManagerPlayerStateChangedNotification
                                                        object:mediaPlayer];
}

#pragma mark - Active Media Player State

- (void)startObservingMediaPlayerChanges
{
    self.titleObserver = [KeyValueObserver observeObject:self.activeMediaPlayer
                                                 keyPath:@"title"
                                                  target:self
                                                selector:@selector(activeMediaPlayerTitleChanged:)
                                                 options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew];
}

- (void)stopObservingMediaPlayerChanges:(id<MediaPlayer>)mediaPlayer
{
    self.titleObserver = nil;
}

- (void)activeMediaPlayerTitleChanged:(NSDictionary *)change
{
    [self.changeObserver activeMediaPlayerTitleDidChange];
}

@end

File: wire-ios/Wire-iOS/Sources/Managers/MediaPlaybackManager.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import avs
import Foundation
import WireSyncEngine
import WireSystem

private let zmLog = ZMSLog(tag: "MediaPlaybackManager")

/// An object that observes changes in the media playback manager.
protocol MediaPlaybackManagerChangeObserver: AnyObject {
    /// The state of the active media player changes.
    func activeMediaPlayerStateDidChange()
}

extension Notification.Name {
    static let mediaPlaybackManagerPlayerStateChanged = Notification.Name("MediaPlaybackManagerPlayerStateChangedNotification")
    static let activeMediaPlayerChanged = Notification.Name("activeMediaPlayerChanged")
}

/// This object is an interface for AVS to control conversation media playback
final class MediaPlaybackManager: NSObject, AVSMedia {

    private let userSession: UserSession
    var audioTrackPlayer: AudioTrackPlayer

    private(set) weak var activeMediaPlayer: MediaPlayer? {
        didSet {
            NotificationCenter.default.post(name: .activeMediaPlayerChanged, object: activeMediaPlayer)
        }
    }

    weak var changeObserver: MediaPlaybackManagerChangeObserver?
    var name: String!

    weak var delegate: AVSMediaDelegate?

    var volume: Float = 0

    var looping: Bool {
        get {
            return false
        }

        set {
            /// no-op
        }
    }

    var playbackMuted: Bool {
        get {
            return false
        }

        set {
            /// no-op
        }
    }

    var recordingMuted: Bool = false

    init(
        name: String?,
        userSession: UserSession
    ) {
        self.userSession = userSession
        self.audioTrackPlayer = AudioTrackPlayer(userSession: userSession)
        super.init()

        self.name = name
        audioTrackPlayer.mediaPlayerDelegate = self
    }

    // MARK: - AVSMedia

    func play() {
        // AUDIO-557 workaround for AVSMediaManager calling play after we say we started to play.
        if activeMediaPlayer?.state != .playing {
            activeMediaPlayer?.play()
        }
    }

    func pause() {
        // AUDIO-557 workaround for AVSMediaManager calling pause after we say we are paused.
        if activeMediaPlayer?.state == .playing {
            activeMediaPlayer?.pause()
        }
    }

    func stop() {
        // AUDIO-557 workaround for AVSMediaManager calling stop after we say we are stopped.
        if activeMediaPlayer?.state != .completed {
            activeMediaPlayer?.stop()
        }
    }

    func resume() {
        activeMediaPlayer?.play()
    }

    func reset() {
        audioTrackPlayer.stop()

        audioTrackPlayer = AudioTrackPlayer(userSession: userSession)
        audioTrackPlayer.mediaPlayerDelegate = self
    }
}

extension MediaPlaybackManager: MediaPlayerDelegate {
    func mediaPlayer(_ mediaPlayer: MediaPlayer, didChangeTo state: MediaPlayerState) {
        zmLog.debug("mediaPlayer changed state: \(state)")

        changeObserver?.activeMediaPlayerStateDidChange()

        switch state {
        case .playing:
            if activeMediaPlayer !== mediaPlayer {
                activeMediaPlayer?.pause()
            }
            delegate?.didStartPlaying(self)
            activeMediaPlayer = mediaPlayer
        case .paused:
            delegate?.didPausePlaying(self)
        case .completed:
            if activeMediaPlayer === mediaPlayer {
                activeMediaPlayer = nil
            }
            delegate?.didFinishPlaying(self) // this interfers with the audio session
        case .error:
            delegate?.didFinishPlaying(self) // this interfers with the audio session
        default:
            break
        }

        NotificationCenter.default.post(name: .mediaPlaybackManagerPlayerStateChanged, object: mediaPlayer)

    }
}
